zebra-script/src/lib.rs

```diff
@@ -65,6 +65,10 @@ pub struct CachedFfiTransaction {
     /// This field is private so that `transaction` and `precomputed` always match.
     transaction: Arc<Transaction>,
 
+    /// The outputs from previous transactions that match each input in the transaction
+    /// being verified.
+    all_previous_outputs: Vec<transparent::Output>,
```

Comment:
Unsafe Constraint Violation:

`Vec` is not `Sync`, but later in this module we:
```rust
unsafe impl Sync for CachedFfiTransaction {}
```

Can you please also update the SAFETY justification on `unsafe impl Sync for CachedFfiTransaction`?

```suggestion
    ///
    /// SAFETY: this field must be private,
    ///         and `CachedFfiTransaction::new` must be the only method that modifies it,
    ///         so that it is [`Send`], [`Sync`], and consistent with `transaction` and `precomputed`.
    all_previous_outputs: Arc<[transparent::Output]>,
```

Reply:
If we want to avoid similar bugs in future refactors, we could make a wrapper type for `precomputed`, and only `unsafe impl Send + Sync` on that wrapper type.

Reply:
I completely missed this, but [`Vec` seems to implement `Sync`](https://doc.rust-lang.org/beta/std/vec/struct.Vec.html#impl-Sync) (since `transparent::Output` also seems to), so I guess we don't need to do anything? Do you think it's best to do the wrapper now?

Reply:
You're right, I didn't check the details of the `Vec` impl and `transparent::Output` impl.

We still need to update the `unsafe impl` docs to explain how `all_previous_outputs` is used with FFI. (Since it's serialized before being used in FFI, the memory used by `all_previous_outputs` is not accessed by FFI at all.)

If we do want to do a wrapper, let's open a separate ticket and schedule it?

Reply:
Added comments in 0ff950eb1

Created https://github.com/ZcashFoundation/zebra/issues/3436